tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,192 @@
+#       ```
+#    - Dimname[]? names
+#       ```cpp
+#           c10::optional<c10::IValue> names_opt = (std::move(peek(stack, 1, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ArrayRef<at::Dimname>> names_opt_out;
+#           if (names_opt.has_value()) {
+#                         ~~~~~~~~~~~ <-- Unwrapping optional shell
+#               const c10::IValue names_opt_in = names_opt.value();
+#               const c10::List<c10::IValue> names_list_in = names_opt_in.toList();
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "c10::IValue {arg_name} = std::move(peek(stack, {pos}, {args_num}));"
+    expr_list = []
+    pop_ivalue = []
+    code_list = []
+    for i, arg in enumerate(args):
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
+        expr, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
```

Comment:
nit: I think that generally the fewer strings you have floating around the codegen, the easier it is to reason about things. Instead of the separate lists for `code_list` and `pop_ivalue`, can you do something like
```
code_list.append(f"""
c10::IValue {arg_name} = std::move(peek(stack, {i}, {len(args)}));
{'\n'.join(block for block in code)}
""")
```

Reply:
The new changes are still pretty non-local. What's nice about the code snippet I put in the above comment are there all of the pieces of code are in the same f-string, so very localized and easy to see why they fit together. ("I create an ivalue named `{arg_name}`, so clearly the next line will use it").